utility_analysis_new/dpengine.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2022 OpenMined.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""DPEngine and ContributionBounder for utility analysis."""
+import collections
+
+import pipeline_dp
+from pipeline_dp import DPEngine
+from pipeline_dp import DataExtractors
+from pipeline_dp import NoiseKind
+from pipeline_dp import contribution_bounders
+from pipeline_dp.contribution_bounders import ContributionBounder
+from dataclasses import dataclass
+
+
+class UtilityAnalysisDPEngine(DPEngine):
+
+    def __init__(self, budget_accountant: 'BudgetAccountant',
+                 backend: 'PipelineBackend'):
+        super().__init__(budget_accountant, backend)
+
+    def aggregate(self,
+                  col,
+                  params: pipeline_dp.AggregateParams,
+                  data_extractors: DataExtractors,
+                  public_partitions=None):
+        return super().aggregate(col, params, data_extractors,
+                                 public_partitions)
+
+    def _aggregate(self, col, params: pipeline_dp.AggregateParams,
+                   data_extractors: DataExtractors, public_partitions):
+        assert params.custom_combiners is None, "Cumtom combiners are not supported"
+        assert params.metrics == [
+            pipeline_dp.Metrics.COUNT
+        ], f"Supported only count metrics, metrics={params.metrics}"
+        assert public_partitions is not None, "Only public partitions supported"
+        assert params.contribution_bounds_already_enforced is False, "Utility Analysis when contribution bounds are already enforced is not supported"
+
+        # TODO: Implement custom combiner
+        combiner = self._create_compound_combiner(params)
+        # Extract the columns.
+        col = self._backend.map(
```

Comment:
it would be ideal to use` _aggregate` from DPEngine with injecting ContributionBounders and Combiners, such that it will be the same computational graph as in production. Would it work?

Reply:
Hmm, I think the partition-selection bit would be problematic. Right now, for utility analysis, public_partitions is technically always set to None (but conceptually we are only doing public partitions), but that'd mean we have to go through the private partition selection logic of _aggregate of DPEngine, which drops some partitions. If we set public_partitions to something other than None, we have to make sure that additional public-partitions-related logic (adding partitions, dropping partitions) do not interfere with the utility analysis.

I don't see an obvious way to make it work without overriding `_aggregate()`. WDYT?

Reply:
no worries about private partition selection for now, I have ideas how to introduce them into `DPEngine.aggregate_` computation graph, I'll need to refine/experiment more with them, but it's doable.

We should use provided `public_partitions` from the user and do normal processing - filtering non public partitions and adding 0-results for missing partitions. And `DPEngine._aggregate` does that (adding of 0-results will be automatic, since `UtilityAnalysisCombiner.create_accumulator `will be called for empty set, we just need to be sure that it returns correct `ErrorPrimitive`.

Does it make sense?